src/box/index.cc

```diff
@@ -336,30 +351,45 @@ box_index_min(uint32_t space_id, uint32_t index_id, const char *key,
 	mp_tuple_assert(key, key_end);
 	struct space *space;
 	struct index *index;
+	const char *key_array = key;
+	uint32_t part_count;
+	struct txn *txn;
+	struct txn_ro_savepoint svp;
+	int rc;
 	if (check_index(space_id, index_id, &space, &index) != 0)
-		return -1;
+		goto error;
 	if (index->def->type != TREE) {
 		/* Show nice error messages in Lua. */
 		diag_set(UnsupportedIndexFeature, index->def, "min()");
-		return -1;
+		goto error;
 	}
-	const char *key_array = key;
-	uint32_t part_count = mp_decode_array(&key);
+	part_count = mp_decode_array(&key);
 	if (key_validate(index->def, ITER_GE, key, part_count))
-		return -1;
+		goto error;
 	box_run_on_select(space, index, ITER_GE, key_array);
 	/* Start transaction in the engine. */
-	struct txn *txn;
-	struct txn_ro_savepoint svp;
 	if (txn_begin_ro_stmt(space, &txn, &svp) != 0)
-		return -1;
-	int rc = index_min(index, key, part_count, result);
+		goto error;
+	rc = index_min(index, key, part_count, result);
 	txn_end_ro_stmt(txn, &svp);
 	if (rc != 0)
-		return -1;
+		goto error;
 	if (*result != NULL)
 		tuple_bless(*result);
 	return 0;
+error:
+	struct error *e = diag_last_error(diag_get());
+	error_set_str(e, "operation", "min");
+	error_set_mp(e, "key", key_array, key_end - key_array);
+	/* space and index pointers can be invalid at this point. */
+	space = space_by_id(space_id);
+	if (space != NULL) {
+		error_set_str(e, "space", space->def->name);
+		index = space_index(space, index_id);
+		if (index != NULL)
+			error_set_str(e, "index", index->def->name);
+	}
```

Comment:
This snippet is repeated a few times. Maybe, factor it out to a helper function?